include/clang/AST/Expr.h

```diff
@@ -4525,36 +4546,57 @@ class BoundsExpr : public Expr {
   SourceLocation getLocStart() const LLVM_READONLY { return StartLoc; }
   SourceLocation getLocEnd() const LLVM_READONLY { return RParenLoc; }
 
+  Kind getKind() const { return (Kind)BoundsExprBits.Kind; }
+  void setKind(Kind Kind) {
+    BoundsExprBits.Kind = Kind;
+    assert(validateKind());
+  }
+
+  bool isInvalid() {
+    return getKind() == Invalid;
+  }
+
+  bool isNone() {
+    return getKind() == None;
+  }
+
+  bool isElementCount() {
+    return getKind() == ElementCount;
+  }
+
+  bool isByteCount() {
+    return getKind() == ByteCount;
+  }
+
+  bool isRange() {
+    return getKind() == Range;
+  }
+
+  bool isPtrInteropAnnotation() {
+    return getKind() == PtrInteropAnnotation;
+  }
+
   static bool classof(const Stmt *T) {
     return T->getStmtClass() >= firstBoundsExprConstant &&
            T->getStmtClass() <= lastBoundsExprConstant;
   }
+
+private:
+  /// \brief: check that the kind is valid for the type
+  /// of bounds expression.
+  bool validateKind();
 };
 
 /// \brief Represents a Checked C nullary bounds expression.
 class NullaryBoundsExpr : public BoundsExpr {
 public:
-  enum Kind {
-    // Invalid bounds expressions are used to flag invalid bounds
-    // expressions and prevent spurious downstream error messages
-    // in bounds declaration checking.
-    Invalid = 0,
-    // bounds(none)
-    None = 1
-  };
-
-public:
   NullaryBoundsExpr(Kind Kind, SourceLocation StartLoc, SourceLocation RParenLoc)
-    : BoundsExpr(NullaryBoundsExprClass, StartLoc, RParenLoc)  {
-    NullaryBoundsExprBits.Kind = Kind;
+    : BoundsExpr(NullaryBoundsExprClass, Kind, StartLoc, RParenLoc)  {
```

Comment:
In these constructors, can we at least add some assertions that the kind indicated is appropriate for the class being constructed?  It otherwise makes me nervous that we'll somewhere accidentally describe a bounds expression that's invalid.